### PR TITLE
Various fixes

### LINKED ISSUE #950: Uncaught Error: ~/repos/myproject/src/components/App.scss does not belong to repo /home/user/repos/myproject/
1. I load a git directory as my current atom project (doesn't have a remote, but has several commits)
2. opened a file that contains unstaged changes
3. tried to view unstaged changes (same thing happens with the staged changes command)
4. got the error

*Note: I've tried this in safe mode as well, same result*

**Atom**: 1.18.0 x64
**Electron**: 1.3.15
**OS**: "Arch Linux"
**Thrown From**: [github](https://github.com/atom/github) package 0.3.3

### Stack Trace

Uncaught Error: /home/adonisk/repos/sideprojects/adonisk.com/src/components/App.scss does not belong to repo /media/adonisk/linux/repos/sideprojects/adonisk.com

```
At /usr/share/atom/resources/app/node_modules/github/lib/controllers/root-controller.js:807

Error: /home/adonisk/repos/sideprojects/adonisk.com/src/components/App.scss does not belong to repo /media/adonisk/linux/repos/sideprojects/adonisk.com
    at RootController.viewChangesForCurrentFile (/usr/share/atom/resources/app/node_modules/github/lib/controllers/root-controller.js:807:19)
    at RootController.viewUnstagedChangesForCurrentFile (/usr/share/atom/resources/app/node_modules/github/lib/controllers/root-controller.js:812:16)
    at CommandRegistry.module.exports.CommandRegistry.handleCommandEvent (/usr/share/atom/resources/app/src/command-registry.js:265:35)
    at CommandRegistry.handleCommandEvent (/usr/share/atom/resources/app/src/command-registry.js:3:65)
    at CommandRegistry.module.exports.CommandRegistry.dispatch (/usr/share/atom/resources/app/src/command-registry.js:166:25)
    at AtomEnvironment.module.exports.AtomEnvironment.dispatchContextMenuCommand (/usr/share/atom/resources/app/src/atom-environment.js:1320:34)
    at EventEmitter.outerCallback (/usr/share/atom/resources/app/src/application-delegate.js:347:31)
    at emitThree (events.js:116:13)
    at EventEmitter.emit (events.js:194:7)
```

### Commands

```
  2x -3:48.6.0 core:backspace (input.hidden-input)
     -3:47.8.0 core:move-down (input.hidden-input)
  2x -3:47.6.0 core:move-up (input.hidden-input)
     -3:47.1.0 docblockr:parse-enter (input.hidden-input)
     -3:47.1.0 editor:newline (input.hidden-input)
     -3:46.9.0 intentions:highlight (input.hidden-input)
     -3:46.8.0 editor:delete-to-beginning-of-word (input.hidden-input)
     -3:46.4.0 intentions:highlight (input.hidden-input)
     -3:46 core:save (input.hidden-input)
     -3:44.1.0 intentions:highlight (input.hidden-input)
  2x -3:44 duplicate-line-or-selection:duplicate (input.hidden-input)
     -3:43.5.0 core:save (input.hidden-input)
     -3:37.5.0 github:open-file (span.github-FilePatchListView-path)
     -3:30.2.0 window:toggle-dev-tools (input.hidden-input)
     -0:10.8.0 github:view-staged-changes-for-current-file (div.line)
     -0:08.4.0 github:view-unstaged-changes-for-current-file (div.line)
```

### Non-Core Packages

```
atom-beautify 0.30.0 
auto-detect-indentation 1.3.0 
autoclose-html 0.23.0 
autocomplete-modules 1.6.10 
ava 0.10.0 
busy-signal 1.4.3 
compare-files 0.8.0 
docblockr 0.10.3 
duotone-dark-syntax 2.1.0 
duplicate-line-or-selection 0.9.0 
editorconfig 2.2.2 
emmet 2.4.3 
file-icons 2.1.7 
file-types 0.5.5 
git-diff-details 1.4.0 
git-time-machine 1.5.9 
gtk-dark-theme 0.1.4 
hidpi 0.12.0 
highlight-selected 0.13.1 
hyperclick 0.0.40 
intentions 1.1.2 
js-hyperclick 1.11.0 
jscs-fixer 1.3.0 
language-apache 1.7.0 
language-babel 2.66.0 
language-dircolors 1.2.0 
language-docker 1.1.8 
language-documentation 2.0.1 
language-graphql 0.9.0 
language-ini 1.19.0 
language-liquid 0.5.1 
language-log 1.8.0 
language-patch 1.2.0 
language-pug 0.0.21 
language-regexp 1.0.1 
language-vcard 0.2.0 
line-diff-details 1.8.0 
linter 2.2.0 
linter-eslint 8.2.1 
linter-flow 5.6.0 
linter-jscs 4.1.2 
linter-jshint 3.1.2 
linter-jsonlint 1.3.0 
linter-ui-default 1.6.1 
live-archive 0.1.15 
local-history 4.3.0 
loved-syntax 1.0.1 
merge-conflicts 1.4.5 
minimap 4.28.2 
minimap-cursorline 0.2.0 
minimap-find-and-replace 4.5.2 
minimap-git-diff 4.3.1 
minimap-highlight-selected 4.6.1 
minimap-linter 2.0.0 
nuclide-format-js 0.0.36 
related 0.3.5 
release-notes 0.53.0 
set-syntax 0.3.2 
sync-settings 0.8.2 
tabulation-control 1.0.8 
tasks 2.6.6 
tool-bar 1.1.0 
tool-bar-almighty 0.8.2 
tree-view-git-status 1.4.0 
vscode-snippets 0.2.0 
```

### LINKED ISSUE #982: Uncaught Error: ~/Documents/ISS247/ISS Framework/swagger/css/style.css does not belong to repo null
[Enter steps to reproduce:]

1. Open any html or js file
2. Right click and View Unstaged Chages. It will thorw error

**Atom**: 1.18.0 x64
**Electron**: 1.3.15
**OS**: Ubuntu 16.04.2
**Thrown From**: [github](https://github.com/atom/github) package 0.3.3

### Stack Trace

Uncaught Error: /home/sanket/Documents/ISS247/ISS Framework/swagger/css/style.css does not belong to repo null

```
At /usr/share/atom/resources/app/node_modules/github/lib/controllers/root-controller.js:807

Error: /home/sanket/Documents/ISS247/ISS Framework/swagger/css/style.css does not belong to repo null
    at RootController.viewChangesForCurrentFile (/usr/share/atom/resources/app/node_modules/github/lib/controllers/root-controller.js:807:19)
    at RootController.viewUnstagedChangesForCurrentFile (/usr/share/atom/resources/app/node_modules/github/lib/controllers/root-controller.js:812:16)
    at CommandRegistry.module.exports.CommandRegistry.handleCommandEvent (/usr/share/atom/resources/app/src/command-registry.js:265:35)
    at CommandRegistry.handleCommandEvent (/usr/share/atom/resources/app/src/command-registry.js:3:65)
    at CommandRegistry.module.exports.CommandRegistry.dispatch (/usr/share/atom/resources/app/src/command-registry.js:166:25)
    at AtomEnvironment.module.exports.AtomEnvironment.dispatchContextMenuCommand (/usr/share/atom/resources/app/src/atom-environment.js:1320:34)
    at EventEmitter.outerCallback (/usr/share/atom/resources/app/src/application-delegate.js:347:31)
    at emitThree (events.js:116:13)
    at EventEmitter.emit (events.js:194:7)
```

### Commands

```
     -4:14.5.0 application:add-project-folder (div.tool-panel.tree-view)
  2x -3:18.8.0 core:close (input.hidden-input)
     -3:15 fuzzy-finder:toggle-file-finder (div.panels-item)
  4x -3:07.0 core:move-down (input.hidden-input)
     -3:05.9.0 core:move-up (input.hidden-input)
     -3:05.4.0 core:confirm (input.hidden-input)
     -2:59.3.0 command-palette:toggle (input.hidden-input)
  2x -2:56.9.0 core:move-down (input.hidden-input)
     -2:56.2.0 core:confirm (input.hidden-input)
     -2:56.2.0 pane:split-left (input.hidden-input)
     -1:45.5.0 core:undo (input.hidden-input)
     -1:23.4.0 core:backspace (input.hidden-input)
     -1:22.2.0 core:move-down (input.hidden-input)
  2x -1:18.5.0 core:undo (input.hidden-input)
  2x -1:06.0 core:move-right (input.hidden-input)
     -0:19.2.0 github:view-unstaged-changes-for-current-file (div.line)
```

### Non-Core Packages

```
pigments 0.39.1 
```

### PR DESCRIPTION
**Pull request by kuychaco**
Fixes #950 - Use real path when checking if file is in repo

Fixes #982 - Handle viewing changes for file not in repo; notify user that repo does not exist, and provide a button to create a new repo. If the user initializes the repo, we retry the original operation to view changes for the opened file.

### CODE REVIEW

File: lib/controllers/root-controller.js [hunk 1]
```diff
@@ -488,8 +497,30 @@ export default class RootController extends React.Component {
...
     const editor = this.props.workspace.getActiveTextEditor();
-    const absFilePath = editor.getPath();
+    const absFilePath = await realPath(editor.getPath());
     const repoPath = this.props.repository.getWorkingDirectoryPath();
+    if (repoPath === null) {
+      const [projectPath] = this.props.project.relativizePath(editor.getPath());
+      const notification = this.props.notificationManager.addInfo(
+        "Hmm, we don't have anything to compare this to.",
+        {
+          description: 'Create a Git repository in order to view changes to the files in your project.',
```
Comment (kuychaco) [inline lib/controllers/root-controller.js]: @jasonrudolph thoughts on how this message aligns with the Atom voice? 💭

File: lib/controllers/root-controller.js [hunk 2]
```diff
@@ -488,8 +497,30 @@ export default class RootController extends React.Component {
 
   async viewChangesForCurrentFile(stagingStatus) {
     const editor = this.props.workspace.getActiveTextEditor();
-    const absFilePath = editor.getPath();
+    const absFilePath = await realPath(editor.getPath());
     const repoPath = this.props.repository.getWorkingDirectoryPath();
+    if (repoPath === null) {
+      const [projectPath] = this.props.project.relativizePath(editor.getPath());
+      const notification = this.props.notificationManager.addInfo(
+        "Hmm, we don't have anything to compare this to.",
```
Comment (jasonrudolph) [inline lib/controllers/root-controller.js]: I :heart: the approachability of this message. The use of "we" reminds me of something @simurai brought up last week. IIRC, he asked whether Atom should ever say "I" or "we?" In other words, can Atom speak in the first person? To me, speaking in the first person gives the impression that Atom is a bot (using "I" or "me") or that Atom is a company or OSS team (using "we"). In my loosely-held opinion, I lean toward having Atom _not_ speak in the first person.

With that in mind, I'd propose something like:

> Hmm, there's nothing to compare this file to

File: lib/controllers/root-controller.js (hunk 1 above)
Comment (jasonrudolph) [inline lib/controllers/root-controller.js]: I wonder if there's more we can do that make it clear that we're presenting this to the user as an _option_, but that they don't have to create a repository if they don't want to. How about something like this?

> You can create a Git repository to track changes to the files in your project.

File: lib/helpers.js [hunk 3]
```diff
@@ -171,11 +171,17 @@ export function getTempDir(prefix = '') {
...
     });
   });
 }
 
+export function realPath(folder) {
+  return new Promise((resolve, reject) => {
+    fs.realpath(folder, (err, rpath) => (err ? reject(err) : resolve(rpath)));
+  });
+}
+
```
Comment (smashwilson) [inline lib/helpers.js]: :+1: Excellent, this will be handy.

As a sidenote - how would we feel about using something like [a recent `fs-extra`](https://www.npmjs.com/package/fs-extra) or [`mz`](https://www.npmjs.com/package/mz) to handle promisified filesystem operations? It'd cut down on the number of helpers we need here. But then, maybe we don't need that many.

File: lib/controllers/root-controller.js [hunk 4]
```diff
@@ -362,13 +367,15 @@ export default class RootController extends React.Component {
   }
 
   @autobind
-  async initializeRepo() {
+  async initializeRepo(initDialogPath) {
     if (this.props.activeWorkingDirectory) {
       await this.acceptInit(this.props.activeWorkingDirectory);
-      return;
+      return null;
```
Comment (smashwilson) [inline lib/controllers/root-controller.js]: :point_up: I think we should ignore the active working directory if an explicit path to initialize is given:

```js
if (!initDialogPath || this.props.activeWorkingDirectory) {
  // ...
}
```

Otherwise I worry that a caller could:

1. Generate the path to initialize
2. Yield the event loop, and something else triggers a `WorkdirContext` change
3. Call `initializeRepo` with the desired path
4. `initializeRepo` will silently init the active working directory instead!

File: lib/controllers/root-controller.js [hunk 5]
```diff
@@ -362,13 +367,15 @@ export default class RootController extends React.Component {
...
     }
 
-    this.setState({initDialogActive: true});
+    return new Promise(resolve => {
+      this.setState({initDialogActive: true, initDialogPath, initDialogResolve: resolve});
+    });
```
Comment (smashwilson) [inline lib/controllers/root-controller.js]: Storing the promise resolution callback in the state like this feels a bit weird to me. I have no cleaner ideas for how to structure this though :grin:

Is it possible for this `setState()` call to be triggered twice before the first resolution callback is triggered... ?

File: lib/controllers/root-controller.js [hunk 6]
```diff
@@ -488,8 +497,30 @@ export default class RootController extends React.Component {
 
   async viewChangesForCurrentFile(stagingStatus) {
     const editor = this.props.workspace.getActiveTextEditor();
-    const absFilePath = editor.getPath();
+    const absFilePath = await realPath(editor.getPath());
```
Comment (smashwilson) [inline lib/controllers/root-controller.js]: :+1:

`realPath` all the things!!!

File: lib/controllers/root-controller.js [hunk 7]
```diff
@@ -488,8 +497,30 @@ export default class RootController extends React.Component {
...
+          dismissable: true,
+          buttons: [{
+            className: 'btn btn-primary',
+            text: 'Create a repository now',
+            onDidClick: async () => {
+              notification.dismiss();
+              const created = await this.initializeRepo(projectPath);
+              // If the user confirmed repository creation,
+              // retry the operation that got them here in the first place
+              if (created) { this.viewChangesForCurrentFile(stagingStatus); }
```
Comment (smashwilson) [inline lib/controllers/root-controller.js]: I think it's possible for the user to delete the suggested project path and initialize a different path instead, in which case the retry would... show this prompt again.

Maybe `initializeRepo` could resolve with the initialized path instead, and the retry could only be done if it's the same as projectPath... ?

File: lib/controllers/root-controller.js (hunk 2 above)
Comment (simurai) [inline lib/controllers/root-controller.js]: @jasonrudolph Thanks for the reminder, I'll see if I can find some more infos around this topic.

File: lib/controllers/root-controller.js (hunk 7 above)
Comment (kuychaco) [inline lib/controllers/root-controller.js]: 👍

File: lib/controllers/root-controller.js (hunk 5 above)
Comment (kuychaco) [inline lib/controllers/root-controller.js]: > Is it possible for this setState() call to be triggered twice before the first resolution callback is triggered... ?

This is a good question and a nice catch. As things are now, I think it's highly unlikely that we run into this issue. We only get into this method from the user clicking the "create repo" button, which opens the dialog, and closing the dialog resolves the promise. There's no opportunity for the user to click the "create repo" button again before the dialog is closed.

The only way I can see this method being called before the resolution is if we add a command for initializing a repo (which seems reasonable). So to address this I'll add a guard at the top of the method and if `this.state.initDialogActive` is already true, we'll do an early return from the method.

File: lib/helpers.js (hunk 3 above)
Comment (kuychaco) [inline lib/helpers.js]: Yeah, not opposed to using a library. But would prefer to not tackle it in this PR

File: lib/controllers/root-controller.js (hunk 4 above)
Comment (kuychaco) [inline lib/controllers/root-controller.js]: Good points. 

After discussion with @BinaryMuse and @smashwilson we decided it’s probably best to drop this conditional and always prompt the user for the directory they want to initialize a repo in. 

Since we’re pre-filling the input for them it’s more like a quick confirmation. And it makes it very explicit where we’ll be initializing the repo. This can be helpful in situations where it’s not quite obvious to the user that we’ll be initializing a repo for the folder that the active item belongs to (say they have multiple panes with active items belonging to different folders)

**Comment by kuychaco** [comment]
Thanks @jasonrudolph and @smashwilson for the reviews and helpful input!

File: lib/controllers/root-controller.js (hunk 2 above)
Comment (simurai) [inline lib/controllers/root-controller.js]: @jasonrudolph Material Design has lots of guidelines around writing messages in products. Here the section about "you" and "we": https://material.io/guidelines/style/writing.html#writing-language

tl;dr: Avoid the pronoun “we”.
